parts/tools/gen_diffuser_frame.py

```
@@ -0,0 +1,848 @@
+                segs.append(dict(kind='rail', pts=[[xr, y0], [xr, y1]],
+                                 chan=ci, gap=len([s for s in segs
+                                                   if s['kind'] == 'rail'
+                                                   and s.get('chan') == ci])))
+            else:
+                print(f'  !! rail x={xr:.2f} blocked between y={y0:.1f}..{y1:.1f}')
+
+    # ---- connectivity bookkeeping ------------------------------------------
+    parent = list(range(n))
+    def uf_find(x):
+        while parent[x] != x:
+            parent[x] = parent[parent[x]]; x = parent[x]
```

Ruff (0.16.0)

[error] Multiple statements on one line (semicolon)

(E702)